axum/src/routing/strip_prefix.rs

```diff
@@ -35,43 +34,402 @@ where
 
     let mut parts = http::uri::Parts::default();
     parts.scheme = uri.scheme().cloned();
     parts.authority = uri.authority().cloned();
     parts.path_and_query = path_and_query;
 
-    Uri::from_parts(parts).unwrap()
+    Some(Uri::from_parts(parts).unwrap())
+}
+
+fn segments(s: &str) -> impl Iterator<Item = &str> {
+    assert!(
+        s.starts_with('/'),
+        "path didn't start with '/'. axum should have caught this higher up."
+    );
+
+    s.split('/')
+}
+
+fn zip_longest<I, I2>(a: I, b: I2) -> impl Iterator<Item = Item<I::Item>>
+where
+    I: Iterator,
+    I2: Iterator<Item = I::Item>,
+{
+    a.map(Some)
+        .chain(std::iter::repeat_with(|| None))
+        .zip(b.map(Some).chain(std::iter::repeat_with(|| None)))
+        .map_while(|(a, b)| match (a, b) {
+            (Some(a), Some(b)) => Some(Item::Both(a, b)),
+            (Some(a), None) => Some(Item::First(a)),
+            (None, Some(b)) => Some(Item::Second(b)),
+            (None, None) => None,
+        })
+}
+
+enum Item<T> {
+    Both(T, T),
+    First(T),
+    Second(T),
+}
+
+#[cfg(test)]
+mod tests {
+    #[allow(unused_imports)]
+    use super::*;
+    use quickcheck::Arbitrary;
+    use quickcheck_macros::quickcheck;
+
+    macro_rules! test {
+        (
+            $name:ident,
+            uri = $uri:literal,
+            prefix = $prefix:literal,
+            expected = $expected:expr $(,)?
+        ) => {
+            #[test]
+            fn $name() {
+                let uri = $uri.parse().unwrap();
+                let new_uri = strip_prefix(&uri, $prefix).map(|uri| uri.to_string());
+                assert_eq!(new_uri.as_deref(), $expected);
+            }
+        };
+    }
+
+    test!(empty, uri = "/", prefix = "/", expected = Some("/"));
+
+    test!(
+        single_segment,
+        uri = "/a",
+        prefix = "/a",
+        expected = Some("/"),
+    );
+
+    test!(
+        single_segment_root_uri,
+        uri = "/",
+        prefix = "/a",
+        expected = None,
+    );
+
+    // the prefix is empty, so removing it should have no effect
+    test!(
+        single_segment_root_prefix,
+        uri = "/a",
+        prefix = "/",
+        expected = None,
+    );
+
+    test!(
+        single_segment_no_match,
+        uri = "/a",
+        prefix = "/b",
+        expected = None,
+    );
+
+    test!(
+        single_segment_trailing_slash,
+        uri = "/a/",
+        prefix = "/a/",
+        expected = Some("/")
+    );
+
+    test!(
+        single_segment_trailing_slash_2,
+        uri = "/a",
+        prefix = "/a/",
+        expected = None,
+    );
+
+    test!(
+        single_segment_trailing_slash_3,
+        uri = "/a/",
+        prefix = "/a",
+        expected = Some("/")
+    );
+
+    test!(
+        multi_segment,
+        uri = "/a/b",
+        prefix = "/a",
+        expected = Some("/b")
+    );
+
+    test!(
+        multi_segment_2,
+        uri = "/b/a",
+        prefix = "/a",
+        expected = None,
```

Comment:
Nitpick, would make sense to be consistent about trailing commas after `expected = …` 😉